documents/coredocument.go

```diff
@@ -197,6 +202,47 @@ func (cd *CoreDocument) PrepareNewVersion(collaborators []string, documentPrefix
 
 }
 
+// PrepareNewVersion prepares the next version of the CoreDocument
+// if initSalts is true, salts will be generated for new version.
+func (cd *CoreDocument) PrepareNewVersion1(newCollaborators Collaborators, documentPrefix []byte) (*CoreDocument, error) {
+	if len(cd.Document.DocumentRoot) != idSize {
+		return nil, errors.New("document root is invalid")
+	}
+
+	// get all the old collaborators
+	oldCs, err := cd.GetCollaborators1()
+	if err != nil {
+		return nil, err
+	}
+
+	rcs := filterCollaborators(newCollaborators.ReadCollaborators, oldCs.ReadCollaborators...)
+	wcs := filterCollaborators(newCollaborators.ReadWriteCollaborators, oldCs.ReadWriteCollaborators...)
+	rcs = append(rcs, wcs...)
+
+	cdp := coredocumentpb.CoreDocument{
+		DocumentIdentifier: cd.Document.DocumentIdentifier,
+		PreviousRoot:       cd.Document.DocumentRoot,
+		Roles:              cd.Document.Roles,
+		ReadRules:          cd.Document.ReadRules,
+		TransitionRules:    cd.Document.TransitionRules,
+		Nfts:               cd.Document.Nfts,
+		AccessTokens:       cd.Document.AccessTokens,
+		SignatureData:      new(coredocumentpb.SignatureData),
+	}
+
+	err = populateVersions(&cdp, &cd.Document)
+	if err != nil {
+		return nil, err
+	}
+
+	ncd := &CoreDocument{Document: cdp}
+	ncd.addCollaboratorsToReadSignRules(rcs)
```

Comment:
Unless `ncd.addCollaboratorsToTransitionRules` as well adds the read rules needed, we need to make this call as well with the write collabs:
`ncd.addCollaboratorsToReadSignRules(wcs)`